theano/gpuarray/c_code/dnn_sptf_gi.c

```diff
@@ -0,0 +1,194 @@
+                             GA_C_ORDER, gpu_ctx ) != 0 )
+        return 1;
+
+    if ( theano_prep_output( grid_grad, PyGpuArray_NDIM( grid ),
+                             PyGpuArray_DIMS( grid ), grid->ga.typecode,
+                             GA_C_ORDER, gpu_ctx ) != 0 )
+        return 1;
+
+    // Obtain output dimensions to setup descriptor
+    out_dims[0] = (int) PyGpuArray_DIM(input, 0); // num_images
+    out_dims[1] = (int) PyGpuArray_DIM(input, 1); // num_channels
+    out_dims[2] = (int) PyGpuArray_DIM(grid, 1); // grid height
+    out_dims[3] = (int) PyGpuArray_DIM(grid, 2); // grid width
+
+    // Currently, only the bilinear sampler is supported by cuDNN,
+    // so the sampler method is currently not available as a parameter
+    err = cudnnSetSpatialTransformerNdDescriptor(APPLY_SPECIFIC(sptf), CUDNN_SAMPLER_BILINEAR,
+        dt, 4, out_dims );
+    if ( CUDNN_STATUS_SUCCESS != err )
+    {
+        PyErr_Format( PyExc_MemoryError,
+            "GpuDnnTransformerGradI: could not initialize descriptor (sptf): %s",
+            cudnnGetErrorString( err ) );
+        return 1;
+    }
+
+    if ( c_set_tensorNd( input, APPLY_SPECIFIC(xdesc) ) != 0 )
+        return 1;
+
+    if ( c_set_tensorNd( dy, APPLY_SPECIFIC(dydesc) ) != 0 )
+        return 1;
+
+    if ( c_set_tensorNd( *input_grad, APPLY_SPECIFIC(dxdesc) ) != 0 )
+        return 1;
+
+    // Directly return the outputs if any of the dimensions is 0.
+    // (cuDNN does not support zero-length dimensions.)
+    if ( PyGpuArray_SIZE( *input_grad ) == 0 || PyGpuArray_SIZE( *grid_grad ) == 0 )
+        return 0;
+
+    cuda_enter( gpu_ctx->ctx );
+
+    cuda_wait( input->ga.data, GPUARRAY_CUDA_WAIT_READ );
+    cuda_wait( grid->ga.data, GPUARRAY_CUDA_WAIT_READ );
+    cuda_wait( dy->ga.data, GPUARRAY_CUDA_WAIT_READ );
+    cuda_wait( (*input_grad)->ga.data, GPUARRAY_CUDA_WAIT_WRITE );
+    cuda_wait( (*grid_grad)->ga.data, GPUARRAY_CUDA_WAIT_WRITE );
+
+    err = cudnnSpatialTfSamplerBackward( _handle, APPLY_SPECIFIC(sptf), alpha_p,
+        APPLY_SPECIFIC(xdesc), PyGpuArray_DEV_DATA( input ), beta_p,
+        APPLY_SPECIFIC(dxdesc), PyGpuArray_DEV_DATA( *input_grad ), alpha_p,
+        APPLY_SPECIFIC(dydesc), PyGpuArray_DEV_DATA( dy ), PyGpuArray_DEV_DATA( grid ),
+        beta_p, PyGpuArray_DEV_DATA( *grid_grad ) );
+
+    cuda_record( input->ga.data, GPUARRAY_CUDA_WAIT_READ );
+    cuda_record( grid->ga.data, GPUARRAY_CUDA_WAIT_READ );
+    cuda_record( dy->ga.data, GPUARRAY_CUDA_WAIT_READ );
+    cuda_record( (*input_grad)->ga.data, GPUARRAY_CUDA_WAIT_WRITE );
+    cuda_record( (*grid_grad)->ga.data, GPUARRAY_CUDA_WAIT_WRITE );
+
+    cuda_exit( gpu_ctx->ctx );
+
+    if ( CUDNN_STATUS_SUCCESS != err )
+    {
+        PyErr_Format( PyExc_RuntimeError,
+            "GpuDnnTransformerGradI: failed to compute gradients of the inputs: %s",
+            cudnnGetErrorString( err ) );
+
+        return -1;
```

Comment:
`return 1` instead here, as for the other error codes returned (just to make code consistent).